src/Features/Core/Portable/CustomMessageHandler/CustomMessageHandlerService.cs

```diff
@@ -0,0 +1,350 @@
+
+                    foreach (var assembly in extension.Assemblies)
+                    {
+                        if (assembly.Value.HasValue)
+                        {
+                            foreach (var handler in assembly.Value.Value.Handlers)
+                            {
+                                _handlers.Remove(handler);
+                            }
+
+                            foreach (var documentHandler in assembly.Value.Value.DocumentHandlers)
+                            {
+                                _documentHandlers.Remove(documentHandler);
+                            }
+                        }
+                    }
+                }
+            }
+
+            extension.AssemblyLoadContext?.Unload();
+        }
+        catch (Exception e) when (LogAndPropagate(e))
+        {
+            // unreachable
+        }
+
+        bool LogAndPropagate(Exception e)
+        {
+            Logger.Log(
+                FunctionId.CustomMessageHandlerService_UnloadCustomMessageHandlerAsync,
+                $"Error unloading {assemblyFolderPath}: {e}",
+                LogLevel.Error);
+            return false;
+        }
+
+        return ValueTask.CompletedTask;
+    }
+
+    public ValueTask ResetAsync(CancellationToken cancellationToken)
+    {
+        Clear();
+        return ValueTask.CompletedTask;
+    }
+
+    public void Dispose()
+        => Clear();
+
+    private void Clear()
+    {
+        List<CustomMessageHandlerExtension> extensions;
+        lock (_lockObject)
+        {
+            extensions = _extensions.Values.ToList();
+            _extensions.Clear();
+            _handlers.Clear();
+            _documentHandlers.Clear();
+        }
+
+        foreach (var extension in extensions)
+        {
+            extension.AssemblyLoadContext.Unload();
+        }
+    }
+
+    private readonly struct CustomMessageHandlerExtension(AssemblyLoadContext assemblyLoadContext) : IEquatable<CustomMessageHandlerExtension>
+    {
+        public AssemblyLoadContext AssemblyLoadContext { get; } = assemblyLoadContext;
+
+        public Dictionary<string, CustomMessageHandlerAssembly?> Assemblies { get; } = new();
+
+        public object AssemblyLoadLockObject { get; } = new();
+
+        // Used to avoid race conditions between RegisterHandlersAsync and UnloadCustomMessageHandlersAsync
+        public bool Equals(CustomMessageHandlerExtension other)
+            => ReferenceEquals(Assemblies, other.Assemblies);
+
+        public override bool Equals([NotNullWhen(true)] object? obj)
+            => obj is CustomMessageHandlerExtension other && Equals(other);
+
+        public override int GetHashCode()
+            => RuntimeHelpers.GetHashCode(Assemblies);
+    }
+
+    private readonly struct CustomMessageHandlerAssembly
+    {
+        public required HashSet<string> DocumentHandlers { get; init; }
+
+        public required HashSet<string> Handlers { get; init; }
```

Comment:
would like docs on what hte strings are.    woudl like docs on what the diff is between DocHandlers and Handlers.    Would like as much immutability as possible.